pygen/pygen_src/riscv_asm_program_gen.py

```diff
@@ -0,0 +1,406 @@
+        # program end
+        self.gen_program_end(hart)
+
+    def gen_kernel_sections(self, hart):
+        pass
+
+    def gen_kernel_program(self, hart, seq):
+        pass
+
+    def gen_sub_program(self, hart, sub_program,
+                        sub_program_name, num_sub_program,
+                        is_debug = 0, prefix = "sub"):
+        pass
+
+    def gen_callstack(self, main_program, sub_program,
+                      sub_program_name, num_sub_program):
+        pass
+
+    def insert_sub_program(self, sub_program, instr_list):
+        pass
+
+    def gen_program_header(self):
+        string = []
+        self.instr_stream.append(".include \"user_define.h\"")
+        self.instr_stream.append(".global_start")
+        self.instr_stream.append(".section .text")
+        if(cfg.disable_compressed_instr):
+            self.instr_stream.append(".option norvc;")
+
+        string.append("csrr x5 mhartid")
+
+        for hart in range(cfg.num_of_harts):
+            string.append("li x6, {}\n{}beq x5, x6, {}f".format(hart, pkg_ins.indent, hart))
+
+        self.gen_section("_start", string)
+
+        for hart in range(cfg.num_of_harts):
+            self.instr_stream.append("{}: j h{}_start".format(hart, hart))
+
+    def gen_program_end(self, hart):
+        if(hart == 0):
+            self.gen_section("write_tohost", ["sw gp", "tohost", "t5"])
+            self.gen_section("_exit", ["j write_tohost"])
+
+    def gen_data_page_begin(self, hart):
+        pass
+
+    def gen_data_page(self, hart, is_kernel = 0, amo = 0):
+        pass
+
+    def gen_stack_section(self, hart):
+        pass
+
+    def gen_kernel_stack_section(self, hart):
+        pass
+
+    def gen_init_section(self, hart):
+        string = pkg_ins.format_string("init:", pkg_ins.LABEL_STR_LEN)
+        self.instr_stream.append(string)
+        self.init_gpr()
+        # Init stack pointer to point to the end of the user stack
+        string = "{}la x{} {}user_stack_end".format(
+            pkg_ins.indent, cfg.sp, pkg_ins.hart_prefix(hart))
+        self.instr_stream.append(string)
+        if (cfg.enable_floating_point):
+            self.init_floating_point_gpr()
+        if (cfg.enable_vector_extension):
+            self.init_vector_engine()
+        self.core_is_initialized()
+        self.gen_dummy_csr_write()
+
+        if (rcs.support_pmp):
+            string = pkg_ins.indent + "j main"
+            self.instr_stream.append(string)
+
+    def setup_misa(self):
+        # TO DO
+        misa = 0b01000000
+        self.instr_stream.append("{}li x{}, {}".format(pkg_ins.indent, cfg.gpr[0], hex(misa)))
+        self.instr_stream.append("{}csrw misa, x{}".format(pkg_ins.indent, cfg.gpr[0]))
+
+    def core_is_initialized(self):
+        pass
+
+    def gen_dummy_csr_write(self):
+        pass
+
+    def init_gpr(self):
+        reg_val = BitArray(uint = 0, length = pkg_ins.DATA_WIDTH)
+        dist_lst = []
+
+        for dist_val in range(5):
```

Comment:
maybe move the `dist_lst.append(reg_val)` outside of the `if...elif...else` block to avoid duplication?
```python
if dist_val == 0:
  ...
...
if dist_val == 3:
  ...
else:
  ...
dist_lst.append(reg_val)
```